# Ornstein-Uhlenbeck Process Simulation

To demonstrate simulation of the Ornstein-Uhlenbeck process, we'll simulate the following process...

$$ dX_t = -2X_tdt + 3dB_t, X_0 = 2, $$

with step $0.01$, for time horizon $T = 1$.

We'll repeat this 1000 times, find empirical mean $\mathbb{E}(X_1)$, and compare this with the true value.

## Exact Solution:

The SDE for the Ornstein-Uhlenbeck process is...

$$ dX_t = -r(X_t - \mu)dt + \sigma B_t $$

Thus, for the Ornstein-Uhlenbeck process defined above, we have $\mu = 0$, $r = 2$, and $\sigma = 3$.

Since $X_0 = 2$ is constant, then by the equation...

$$ \int_{0}^{t} g(s) dB_s \sim \text{Normal} \left( 0, \int_{0}^{t} g^2(s) ds \right), $$

we know...

$$ \mathbb{E}(X_t) = \mu + (X_0 - \mu)e^{-rt}. $$

Thus, our exact expectation for $t = 1$ is calculated to be...

$$ \mathbb{E}(X_1) = 0 + (2 - 0)e^{-2*1} = 2e^{-2} = 0.2707 $$

## Simulation:

To approximate the SDE...

$$ dX_t = -2X_tdt + 3dB_t \text{ for } X_0 = 2,$$

we will use the approximation...

$$ X_{k+1} = X_k - r\frac{T}{n}(X_k - \mu) + \sigma Z_i \sqrt{T/n} \text{ for } k = 0, 1, \dots, n-1. $$

For $\mu = 0$, $r = 2$, $\sigma = 3$, $T = 1$, and $n = 100$, this simplifies to...

$$ X_{k+1} = 0.98X_k + 0.3Z_k \text{ for } k = 0, 1, \dots, n-1. $$

Solving the recursion, we find...

$$ X_{n} = 0.98^{n} X_0 + \sum_{i=0}^{n-1} \left[ 0.3Z_i(0.98)^{n-i-1} \right] \text{ for } n = 100.$$

This method generations a discretized approximate sample path for $(X_t)_{0 \leq t \leq T}$, from which we will calculate the empirical value of $\mathbb{E}(X_1)$.

In [1]:
import numpy as np
import math
import matplotlib.pyplot as plt

In [2]:
# t = end of interval in [0,t]
# n = number of increments between 0 and t
# N = number of simulations

def ornUhlenApprox():
    X0 = 2
    Z_array = np.random.normal(0,1,100)
    Zsum = sum([0.3*Z_array[i]*((0.98)**(99-i)) for i in range(100)])
    Xn = (0.98**100)*X0 + Zsum
    return Xn

def ornUhlenSim(N):
    np.random.seed(0)
    ornUhlenApproxs = [ornUhlenApprox() for i in range(N)]
    empMean = np.mean(ornUhlenApproxs)
    return empMean

ornUhlenSim(1000)

0.2657431064555142

As seen in the simulation above, our empirical mean of the SDE of the given Ornstein-Uhlenbeck process is about $0.2657$. We observe that this value is very close to our calculated expectation, $\mathbb{E}(X_1) = 0.2707$.